File: Java/Fish.cpp
#include "Fish.hpp"
#include "Aquarium.hpp"
#include <cmath>
#define PI 3.14159265

//Abstract Base Class

//Berenang ke suatu titik (to_x,to_y) dengan kecepatan sebesar speed
void Fish :: swimto(Point x, int speed){
	Point temp = getPosisi();
	double rad = atan2(x.getOrdinat() - temp.getOrdinat() ,
					x.getAbsis() - temp.getAbsis());
	int deg = (rad * 180) /PI;
	swim(deg,speed);
}

void Fish :: swim(int degree, int speed){
	double rad = (double) degree * PI / 180.0;
	Point temp = getPosisi();
	double absis_new = temp.getAbsis()+(speed*cos(rad)*((MAX_MOVE_TIME)-getMoveTime())/15);
	double ordinat_new = temp.getOrdinat()+(speed*sin(rad)*(MAX_MOVE_TIME-getMoveTime())/15); 
	if(absis_new<temp.getAbsis()){
		right_direct=false;
	} else{
		right_direct=true;
	}
	double absis = absis_new;
	double ordinat = ordinat_new;

	if (absis <= 30) absis = 30;
	if (absis >= 870) absis = 870;
	if (ordinat <= 50) ordinat = 50;
	if (ordinat >= 600) ordinat = 600;

	temp.setAbsis(absis);
	temp.setOrdinat(ordinat);

	setPosisi(temp);
}

//Change Direction
void Fish :: change_direction(){
	right_direct = !right_direct;
}

//Cek lapar
bool Fish :: isStarving(){
	return starvation_period <= 0;
}

//Cek arah
bool Fish :: isFaceRight(){
	return right_direct;
}

//Setter Getter
int Fish :: getMoveTime(){
	return move_timer;
}
int Fish :: getID(){
	return id;
}
Point Fish :: getPosisi(){
	return posisi;
}
int Fish :: getStarvationPeriod(){
	return starvation_period;
}
int Fish :: getSpeed(){
	return speed;
}
int Fish :: getStarving(){
	return starving;
}
int Fish :: getDegree(){
	return degrees;
}
double Fish :: getRadius(){
	return radius;
}
string Fish :: getJenis(){
	return jenis;
}
int Fish :: getHarga(){
	return harga;
}
int Fish :: getMAXMOVE(){
	return MAX_MOVE_TIME;
}
void Fish :: setMoveTime(int move_time){
	move_timer = move_time;
}
void Fish :: setRadius(double x){
	radius = x;
}
void Fish :: setJenis(string s){
	jenis = s;
}
void Fish :: setPosisi(Point x){
	posisi = x;
}

void Fish::setStarvationPeriod(int newStarvationPeriod){
	starvation_period = newStarvationPeriod;
}

void Fish :: setStarving(int new_starving){
	starving = new_starving;
}

void Fish :: setHarga(int harga){
	this->harga = harga;
}
void Fish :: setDegree(int deg){
	degrees = deg;
}

File: Java/Fish.java
import java.lang.Math;

public abstract class Fish implements Comparable {

  private final double pi = 3.14159265;
  private int maxStarving;
  private int maxStarvationPeriod;
  private int maxMoveTime;
  private static int counter = 0;
  private int id;
  private boolean rightDirect;
  private int starvationPeriod;
  private int speed;
  private int starving;
  private int harga;
  private int degree;
  private Point posisi;
  private String jenis;

  public abstract void eat(Aquarium aquarium);

  public abstract void produce(Aquarium aquarium);

  public abstract int getProduceTime();

  public abstract int getMaxProduceTime();

  public abstract void setProduceTime(int produceTime);

  public void swimto(Point x, int speed) {
    Point temp = this.getPosisi();
    double rad = atan2((double)(x.getOrdinat() - temp.getOrdinat()),
        (double)(x.getAbsis() - temp.getAbsis()));
    int deg = (rad * 180) / pi;
    swim(deg, speed);
  }

  public void swim(int degree, int speed) {
    double rad = (double) degree * pi / 180.0;
    Point temp = getPosisi();
    int absisNew =
        temp.getAbsis() + (speed * cos(rad) * ((maxMoveTime) - getMoveTime()) / 15);
    int ordinatNew =
        temp.getOrdinat() + (speed * sin(rad) * (maxMoveTime - getMoveTime()) / 15);
    if (absisNew < temp.getAbsis()) {
      rightDirect = false;
    } else {
      rightDirect = true;
    }
    int absis = absisNew;
    int ordinat = ordinatNew;

    if (absis <= 30) {
      absis = 30;
    }
    if (absis >= 870) {
      absis = 870;
    }
    if (ordinat <= 50) {
      ordinat = 50;
    }
    if (ordinat >= 600) {
      ordinat = 600;
    }

    temp.setAbsis(absis);
    temp.setOrdinat(ordinat);

    setPosisi(temp);
  }

  //Change Direction
  public void change_direction() {
    rightDirect = !rightDirect;
  }

  @Override
  public int compareTo(Object object) {
    Fish fish = (Fish) object;
    if (fish.getId() == getId()) {
      return 0;
    } else if (fish.getId() < getId()) {
      return -1;
    } else {
      return 1;
    }
  }

  //Cek lapar
  public boolean isStarving() {
    return starvationPeriod <= 0;
  }

  //Cek arah
  public boolean isFaceRight() {
    return rightDirect;
  }

  //Setter Getter
  public int getMoveTime() {
    return moveTimer;
  }

  public int getId() {
    return id;
  }

  public Point getPosisi() {
    return posisi;
  }

  public int getStarvationPeriod() {
    return starvationPeriod;
  }

  public int getSpeed() {
    return speed;
  }

  public int getHarga() {
    return harga;
  }

  public int getStarving() {
    return starving;
  }

  public int getDegree() {
    return degrees;
  }

  public int getMaxMove() {
    return maxMoveTime;
  }

  public double getRadius() {
    return radius;
  }

  public String getJenis() {
    return jenis;
  }

  public void setDegree(int degree) {
    degree = degree;
  }

  public void setMoveTime(int moveTimer) {
    this.moveTimer = moveTimer;
  }

  public void setHarga(int harga) {
    this.harga = harga;
  }

  public void setRadius(double radius) {
    this.radius = radius;
  }

  public void setJenis(String jenis) {
    this.jenis = jenis;
  }

  public void setPosisi(Point posisi) {
    this.posisi = posisi;
  }

  public void setStarvationPeriod(int starvationPeriod) {
    this.starvationPeriod = starvationPeriod;
  }

  public void setStarving(int starving) {
    this.starving = starving;
  }
}